### PR TITLE
ARROW-17966: [C++] Adjust to new format for Substrait optional arguments

### PR DESCRIPTION
**Pull request by westonpace**

### CODE REVIEW

**Comment by westonpace** [comment]
This PR currently depends on a non-official (PR fork) version of Substrait.  We probably should not merge until the corresponding Substrait PR has merged.

File: cpp/src/arrow/engine/substrait/expression_internal.cc
Comment (bkietz) [inline cpp/src/arrow/engine/substrait/expression_internal.cc:99]:
```suggestion
  for (const auto& option : scalar_fn.options()) {
    DecodeOption(option, &call);
```

File: cpp/src/arrow/engine/substrait/plan_internal.cc
Comment (bkietz) [inline cpp/src/arrow/engine/substrait/plan_internal.cc:138]: There does not appear to be a way to inspect a substrait clone and get the version numbers apart from parsing CHANGELOG.md. You could add `#cmakedefine ARROW_SUBSTRAIT_BUILD_VERSION` to config.h.cmake then parse that here, or parse it in cmake and add `#cmakedefine ARROW_SUBSTRAIT_BUILD_VERSION_MAJOR` etc to config.h.cmake. This is for example how ThirdpartyToolchain.cmake provides a `#define` to indicate that jemalloc is vendored https://github.com/drin/arrow/blob/dea465396f981f9bd9862e501dc6750ca4fee6d9/cpp/cmake_modules/ThirdpartyToolchain.cmake#L1864-L1865 -> https://github.com/drin/arrow/blob/dea465396f981f9bd9862e501dc6750ca4fee6d9/cpp/src/arrow/util/config.h.cmake#L48

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/engine/substrait/extension_set.cc
Comment (bkietz) [inline cpp/src/arrow/engine/substrait/extension_set.cc:178]: Should this assert `!option_preferences.empty()`?

File: cpp/src/arrow/engine/substrait/extension_set.cc
Comment (bkietz) [inline cpp/src/arrow/engine/substrait/extension_set.cc:672]: It's probably not necessary to reify "UNSPECIFIED" in all the enumerations. We will always either coerce to a default or error because a real value was required. That could be expressed in EnumParser as:
```c++
template <typename Enum>
using EnumParser = std::function<Result<Enum>(std::string_view repr, std::optional<Enum> default)>;
```
... and it would save you some one-based indexing headache.

Probably out of scope for this PR

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/engine/substrait/extension_set.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/extension_set.cc:183]: Ah, true, could have done a `std::move` on `prefs_copy`.  I agree this is more readable though.

File: cpp/src/arrow/engine/substrait/extension_set.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/extension_set.cc:717]: Done, and added a test case for this situation.

File: cpp/src/arrow/engine/substrait/extension_set.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/extension_set.cc]: I made it `ParseOptionOrElse` just to be contrarian.

File: cpp/src/arrow/engine/substrait/extension_set.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/extension_set.cc:179]: I went ahead and made it a full-blown invalid status as I can see this happening and it might be nice to be graceful about it.

File: cpp/src/arrow/engine/substrait/expression_internal.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/expression_internal.cc]: Fixed.

**Comment by westonpace** [comment]
@bkietz Thanks for the review.  I still need to take a look at the cmake changes but I'll try that out later.

File: cpp/src/arrow/engine/substrait/extension_set.cc
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/extension_set.cc]: Agreed to leave out of scope for this PR.  I also agree it would probably be a good idea.  The "unspecified" thing is a relic from the old behavior where "unspecified" was something you had to explicitly specify.

**Comment by westonpace** [comment]
Deferring the cmake changes to https://issues.apache.org/jira/browse/ARROW-18145

**Comment by westonpace** [comment]
@bkietz I'm moving this to ready-for-review.  @rtpsw has been trying to get Ibis / Arrow compatibility working and this PR enables that, even though it temporarily breaks us from the main branch of Substrait while we wait on https://github.com/substrait-io/substrait/pull/342 to merge.  However, there is some work that is stacking up on top of this and branch management has started to cause some headaches.

Curious to get your thoughts on moving forward as-is with this PR, to make it easier for people working on integration, and addressing any remaining discrepancies (e.g. Enum vs string) in follow-up PRs.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (bkietz) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:660]: Moving forward sounds fine to me; this is why we're pulling this dependency from a URL after all. I've added https://issues.apache.org/jira/browse/ARROW-18201 to track the reversion. In the meantime, please move this modification to [versions.txt](https://github.com/westonpace/arrow/blob/cdb16492027dec59e6b8c55440077bd0dd3c919d/cpp/thirdparty/versions.txt#L84-L85), annotated with `// TODO(ARROW-18201)`

**Review by bkietz — CHANGES_REQUESTED** [review]

**Comment by westonpace** [comment]
The feature has now merged into Substrait.  Unfortunately, it seems releases are automated on Sundays so this probably can't merge until after Sunday at which time I will change the version numbers to be 0.20.0.

File: cpp/src/arrow/engine/substrait/expression_internal.cc
Comment (bkietz) [inline cpp/src/arrow/engine/substrait/expression_internal.cc:81]:
```suggestion
  call->SetOption(opt.name(), std::move(prefs));
```

**Review by bkietz — CHANGES_REQUESTED** [review]

**Comment by westonpace** [comment]
@rok @bkietz Now that substrait version 0.20.0 is out, with this change, I think this is good to go.

**Pull request merged**
